File: 2022/06/solution.py
def parse_input():
    with open('input.txt') as f:
        input = f.read()
    return input.strip()


def part1(input):
    for i in range(len(input) - 4):
        window = set(input[i: i + 4])
        if len(window) == 4:
            return i + 4


def part2(input):
    for i in range(len(input) - 14):
        window = set(input[i: i + 14])
        if len(window) == 14:
            return i + 14


input = parse_input()
print(part1(input))
print(part2(input))

File: 2022/06/solution.rs
use std::collections::HashSet;
use std::fs;
use std::iter::FromIterator;

fn main() {
    let input = parse_input();
    println!("{}", part1(&input));
    println!("{}", part2(&input));
}

fn parse_input() -> Vec<char> {
    fs::read_to_string("input.txt").unwrap().chars().collect()
}

fn part1(input: &[char]) -> usize {
    input
        .windows(4)
        .enumerate()
        .find(|(_, window)| HashSet::<char>::from_iter(window.iter().copied()).len() == 4)
        .unwrap()
        .0
        + 4
}

fn part2(input: &[char]) -> usize {
    input
        .windows(14)
        .enumerate()
        .find(|(_, window)| HashSet::<char>::from_iter(window.iter().copied()).len() == 14)
        .unwrap()
        .0
        + 14
}
